File: y2023/day8/day8.py
import math

content = open('y2023/day8/input.txt').read()

def get_stuff() -> tuple[str, dict[str, tuple[str, str]]]:
    [dirs, lookup_str] = content.split('\n\n', 1)
    dirs = dirs.strip()
    lookup = {}
    for lookup_ln in lookup_str.split('\n'):
        for c in '()=,':
            lookup_ln = lookup_ln.replace(c, '')
        [key, left, right] = lookup_ln.split(maxsplit=2)
        lookup[key] = (left, right)
    return (dirs, lookup)

dir_map = {
    'L': 0,
    'R': 1,
}

def part_1():
    dirs, lookup = get_stuff()
    current = 'AAA'
    steps = 0
    while current != 'ZZZ':
        current_dir = dirs[steps % len(dirs)]
        current = lookup[current][dir_map[current_dir]]
        steps += 1
    print(steps)

def part_2_helper(start: 'str', dirs: str, lookup: dict[str, tuple[str, str]]) -> int:
    steps = 0
    current = start
    while current[-1] != 'Z':
        current_dir = dirs[steps % len(dirs)]
        current = lookup[current][dir_map[current_dir]]
        steps += 1
    return(steps)

def part_2():
    step_ls = []
    dirs, lookup = get_stuff()
    for loc in lookup:
        if loc[-1] == 'A':
            step_ls.append(part_2_helper(loc, dirs, lookup))
    print(math.lcm(*step_ls))

if __name__ == '__main__':
    part_2()

File: y2023/day8/day8.exs
defmodule BasicMath do
  def gcd(a, 0), do: a
  def gcd(0, b), do: b
  def gcd(a, b), do: gcd(b, rem(a, b))

  def lcm(0, 0), do: 0
  def lcm(a, b), do: a * b / gcd(a, b)
end

defmodule Day7 do

  def get_stuff do
    content = File.read("y2023/day8/input.txt") |> elem(1)
    [dirs, lookup_str] = String.split(content, "\n\n", parts: 2)

    lookup = Map.new()

    for lookup_ln <- String.split(lookup_str, "\n") do
      lookup_ln = String.replace(lookup_ln, ~r"[\(\)\=\,]", "")
      [key, left, right] = String.split(lookup_ln, parts: 3)
      lookup = Map.put(lookup, key, {left, right})
    end

    {dirs, lookup}

  end
end

Day7.get_stuff
